staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,620 @@
+// 3) Reject current request if there is not enough concurrency shares and
+// we are at max queue length
+// 4) If not rejected, create a request and enqueue
+// returns the enqueud request on a successful enqueue
+// returns nil in the case that there is no available concurrency or
+// the queuelengthlimit has been reached
+// NOTE: This function must only be called with the QueueSet locked
+func (qs *queueSet) timeoutOldRequestsAndRejectOrEnqueue(hashValue uint64) *fq.Request {
+
+	//	Start with the shuffle sharding, to pick a queue.
+	queueIdx := qs.chooseQueueIndexLocked(hashValue)
+	queue := qs.queues[queueIdx]
+	// The next step is the logic to reject requests that have been waiting too long
+	qs.removeTimedOutRequestsFromQueue(queue)
+	// NOTE: currently timeout is only checked for each new request.  This means that there can be
+	// requests that are in the queue longer than the timeout if there are no new requests
+	// We think this is a fine tradeoff
+
+	// Create a request and enqueue
+	req := &fq.Request{
+		DequeueChannel:  make(chan bool, 1),
+		RealEnqueueTime: qs.clock.Now(),
+		Queue:           queue,
+	}
+	if ok := qs.rejectOrEnqueue(req); !ok {
+		return nil
+	}
+	metrics.ObserveQueueLength(qs.name, len(queue.Requests))
+	return req
+}
+
+// removeTimedOutRequestsFromQueue rejects old requests that have been enqueued
+// past the requestWaitLimit
+func (qs *queueSet) removeTimedOutRequestsFromQueue(queue *fq.Queue) {
+	timeoutIdx := -1
+	now := qs.clock.Now()
+	reqs := queue.Requests
+	// reqs are sorted oldest -> newest
+	// can short circuit loop (break) if oldest packets are not timing out
+	// as newer packets also will not have timed out
+
+	// now - requestWaitLimit = waitLimit
+	waitLimit := now.Add(-qs.requestWaitLimit)
+	for i, req := range reqs {
+		if waitLimit.After(req.RealEnqueueTime) {
+			qs.counter.Add(1)
+			req.DequeueChannel <- false
+			close(req.DequeueChannel)
+			// get idx for timed out packets
+			timeoutIdx = i
+		} else {
+			break
+		}
+	}
+	// remove timed out packets from queue
+	if timeoutIdx != -1 {
+		// timeoutIdx + 1 to remove the last timeout req
+		removeIdx := timeoutIdx + 1
+		// remove all the timeout packets
+		queue.Requests = reqs[removeIdx:]
+		// decrement the # of requestsEnqueued
+		qs.numRequestsEnqueued -= removeIdx
+	}
+}
+
+// getRequestsExecutingLocked gets the # of requests which are "executing":
+// this is the# of requests/packets which have been dequeued but have not had
+// finished (via the FinishRequest method invoked after service)
+func (qs *queueSet) getRequestsExecutingLocked() int {
+	total := 0
+	for _, queue := range qs.queues {
+		total += queue.RequestsExecuting
+	}
+	return total
+}
+
+// chooseQueueIndexLocked uses shuffle sharding to select a queue index
+// using a 'hashValue'.  The 'hashValue' derives a hand from a set range of
+// indexes (range 'desiredNumQueues') and returns the queue with the least queued packets
```

Comment:
There are no `desiredNumQueues` and `handSize` parameters here.  Let's just delete this sentence, and maybe extend the first sentence to say "... using the given hashValue and the shuffle sharding parameters of the queueSet".

Reply:
Updated the comment to read:
```
// chooseQueueIndexLocked uses shuffle sharding to select a queue index
// using the given hashValue and the shuffle sharding parameters of the queueSet.
```